org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java

```diff
@@ -215,6 +227,104 @@ public CodeActionHandler(PreferenceManager preferenceManager) {
+		}
+		final int startPos = start;
+		final int endPos = end;
+		try {
+			final JavaClientConnection connection = JavaLanguageServerPlugin.getInstance().getClientConnection();
+			List<IProblemLocationCore> locations = new ArrayList<>();
+			if (connection != null) {
+				final BaseDiagnosticsHandler handler = new DiagnosticsHandler(connection, unit) {
+
+					@Override
+					public void acceptProblem(IProblem problem) {
+						if (!isSyntaxMode() || isSyntaxLikeError(problem) && !monitor.isCanceled()) {
+							int endProblem = problem.getSourceEnd();
+							int startProblem = problem.getSourceStart();
+							try {
+								if (problem.getID() == IProblem.UndefinedType) {
+									if (unit.getBuffer() != null) {
+										int s = problem.getSourceStart();
+										if (s > 0) {
+											s--;
+											char ch = unit.getBuffer().getChar(s);
+											while (Character.isWhitespace(ch)) {
+												s--;
+												ch = unit.getBuffer().getChar(s);
+											}
+											if (ch == '@') {
+												startProblem = s;
+											}
+										}
+									}
+								}
+							} catch (JavaModelException e) {
+								JavaLanguageServerPlugin.logException(e.getMessage(), e);
+							}
+							if (startProblem >= startPos && endProblem <= endPos) {
+								int problemId = problem.getID();
+								boolean isError = problem.isError();
+								ProblemLocationCore location = new ProblemLocationCore(startProblem, endProblem - startProblem, problemId, new String[0], isError, IJavaModelMarker.JAVA_MODEL_PROBLEM_MARKER);
+								locations.add(location);
+							}
+						}
+					}
+
+					@Override
+					public void endReporting() {
+						JavaLanguageServerPlugin.logInfo(locations.size() + " locations reported for " + this.uri.substring(this.uri.lastIndexOf('/')));
+					}
+
+				};
+				WorkingCopyOwner wcOwner = new WorkingCopyOwner() {
+
+					/* (non-Javadoc)
+					 * @see org.eclipse.jdt.core.WorkingCopyOwner#createBuffer(org.eclipse.jdt.core.ICompilationUnit)
+					 */
+					@Override
+					public IBuffer createBuffer(ICompilationUnit workingCopy) {
+						if (!monitor.isCanceled()) {
+							ICompilationUnit original = workingCopy.getPrimary();
+							IResource resource = original.getResource();
+							if (resource instanceof IFile) {
+								return new DocumentAdapter(workingCopy, (IFile) resource);
+							}
+						}
+						return DocumentAdapter.Null;
+					}
+
+					/* (non-Javadoc)
+					 * @see org.eclipse.jdt.core.WorkingCopyOwner#getProblemRequestor(org.eclipse.jdt.core.ICompilationUnit)
+					 */
+					@Override
+					public IProblemRequestor getProblemRequestor(ICompilationUnit workingCopy) {
+						return handler;
+					}
+
+				};
+				if (!monitor.isCanceled()) {
+					int flags = ICompilationUnit.FORCE_PROBLEM_DETECTION | ICompilationUnit.ENABLE_BINDINGS_RECOVERY | ICompilationUnit.ENABLE_STATEMENTS_RECOVERY;
+					unit.reconcile(ICompilationUnit.NO_AST, flags, wcOwner, new NullProgressMonitor());
```

Comment:
Just wondering, why do we need to reconcile the document here at all ? I figured if the client sends the filtered diagnostics whose code actions we should produce, what else is there to do here ?

Reply:
I have added it if a client doesn't sends the filtered diagnostics - showQuickFixesAtTheLineSupport=false - https://github.com/redhat-developer/vscode-java/pull/2166/commits/f53c0cec48aec145dc5e1778c59d6c72d0731452?authenticity_token=Z98yJTYcuIo0NAKhwD%2BmBCiUSLlW%2Bvz%2F6t3bxigtMPMN9y50ZCU%2FfaxUqtjt5H2BeauGj3BGAuq80A5nnFbNSA%3D%3D&file-filters%5B%5D=.ts#diff-04bba6a35cad1c794cbbe677678a51de13441b7a6ee8592b7b50be1f05c6f626R196

Reply:
+1
Using reconcile in code action is not a good idea. If the client doesn't send the diagnostics, we can disable this feature.

Reply:
It is active only if the client set java.quickfix.showAll=true and doesn't support the diagnostics 
https://github.com/eclipse/eclipse.jdt.ls/pull/1908/commits/79e125a1361278d9ded43de657a3561fc9a7efe6#diff-419f075b7b562e824b4f38f2b6be9714c669ba4e72785216d8352f9517448fdbR231
. 
> if (!this.preferenceManager.getPreferences().isJavaQuickFixShowAll() || this.preferenceManager.getClientPreferences().showQuickFixesAtTheLineSupport()) {

The VS Code client never call it. I can exclude it.

Reply:
I have removed it.

Reply:
Sorry for the confusion here, but I think we may have to restore the logic on the server side for clients that don't support this. For what its worth, clients that do support `showQuickFixesAtTheLineSupport` would never run the code that makes the extra call to reconcile, so they wouldn't feel a performance hit.

If @testforstephen is fine with re-introducing the code, then I would just try to reduce some of the duplication in the logic. It looked similar to what I saw in some base diagnostics handler so maybe some of it can be placed into a helper method ?

Reply:
> Sorry for the confusion here, but I think we may have to restore the logic

@rgrunber I have restored it.

Reply:
>  I would just try to reduce some of the duplication in the logic. It looked similar to what I saw in some base diagnostics handler so maybe some of it can be placed into a helper method ?

I have updated the PR. Could you, please, review it again?

Reply:
> Sorry for the confusion here, but I think we may have to restore the logic on the server side for clients that don't support this. For what its worth, clients that do support showQuickFixesAtTheLineSupport would never run the code that makes the extra call to reconcile, so they wouldn't feel a performance hit.

Since it is cheaper to have the client return all the diagnostics at the line, I would rather ask the client to do so if they want this feature.

Server-side reconcile is not a cheap operation, and I would be more cautious about adding some expensive calculations to a code action, since code action is triggered very frequently. If you want to keep the server code and let other clients get it for free, can you pls do some performance analysis first?

WDYT?

Reply:
`CompilationUnit.reconcile(..)` is the large majority of `CodeActionHandler.getCodeActionCommands(..)` when the client lacks support. It's difficult to get some consistent numbers though since even for large files, it doesn't seem to be very noticeable.

In the worst case I'd say it's probably around double (~100ms vs ~200ms). This is the case of having unsaved changes in a file (which should be more expensive for `reconcile`) and triggering code actions.

Reply:
I also ran cpu profiling against this PR. What i did was to open a Java file with about 700 LOC, then move the cursor up and down a few times and wait for the lightbulb appearing. I didn't do any other editing during this time. See the profiling result below. It shows that the new reconcile logic will give CodeActionHandler an additional 32% CPU time. 700 LOC is a medium line of code for me, and the validation overhead is quite obvious.

![image](https://user-images.githubusercontent.com/14052197/140074980-d507aff6-06fb-401c-8b53-44c05b2e58b2.png)

Reply:
this data is matched with the CPU profiling, it brings an additional 1/3 overhead.

If the reconcile logic is not needed, it's better to remove it to keep the code clean.

Reply:
@rgrunber @fbricon Do you agree?

Reply:
Yes, let's leave it out for now. Is there no other way to get the problems without calling reconcile and including a problem requestor ?

Reply:
There isn't another way.